### DISCUSSION TITLE
HPOS tables and the Polylang/Hyyan WooCommerce Polylang Integration plugins

### DISCUSSION THREAD

**Discussion by mrleemon** [woocommerce/woocommerce#34829] [WooCommerce HPOS]
I tried the HPOS tables in a site with about 1500 completed orders using the multilanguage plugin [Polylang](https://wordpress.org/plugins/polylang/) and the [Hyyan WooCommerce Polylang Integration](https://wordpress.org/plugins/woo-poly-integration/) companion plugin and when I visit the orders admin screen I see no orders and if I run the `wp wc cot sync` command, it says that “There are 0 orders to be synced”.

I think that the relevant code to be adapted to HPOS tables can be found here:
https://github.com/hyyan/woo-poly-integration/blob/master/src/Hyyan/WPI/Order.php

This class basically saves the language in which an order was made ~~in post meta~~ using a `language` taxonomy.

**Comment by vedanshujain** [comment]
Hey @mrleemon , can you share a bit more details (screenshot of the settings page by going to WooCommerce > Settings > Advanced > Custom Data Stores. You should be able to see orders to sync even if there are incompatible plugins unless the plugin is changing `post_type` itself (or the result of `WC_Query`).

**Reply by mrleemon** [reply]
Here it is...

![Screenshot 2022-09-26 at 09-51-18 WooCommerce settings - WordPress](https://user-images.githubusercontent.com/2060659/192222678-60a4e479-4790-4f07-90e1-0d40904f0281.png)

I changed the "Data store for orders" setting back to "Use the WordPress posts table" and then to "Use the WooCommerce orders tables" again, and now when I visit the orders admin screen I get the following fatal error:

![fatal](https://user-images.githubusercontent.com/2060659/192225811-8c5120fd-786d-4ff3-89a4-da196606b377.jpg)

**Reply by vedanshujain** [reply]
It looks like a debug code made it into the testing package, which allows switching data stores as needed (and not showing pending orders).  I have scheduled another build here, it should be ready in 15-20 mins https://github.com/woocommerce/woocommerce/actions/runs/3126279744.

**Comment by mrleemon** [comment]
I tried the 7.0.0-hpos.dev.2 build and I'm still getting the above fatal error when visiting the orders admin screen. I checked the `wp_wc_orders` table via Adminer and is full of data, so apparently the syncing is working.

**Reply by vedanshujain** [reply]
I can repro this now, taking a look. Thanks for reporting!

**Reply by mrleemon** [reply]
If you are testing, I recommend using the master version of the Hyyan WooCommerce Polylang Integration plugin found on GitHub because it has some fixes that the repository version doesn't have:

https://github.com/hyyan/woo-poly-integration

**Reply by vedanshujain** [reply]
I was able to repro with just core, actually, the issue was happening because of an incorrect read operation when syncing between order tables and wp_posts. [Added a fix](https://github.com/woocommerce/woocommerce/commit/c8069b040b68c5828be7fbb1ea121431499f294c) and started a new [build](https://github.com/woocommerce/woocommerce/actions/runs/3127710141), should be ready soon.

**Reply by mrleemon** [reply]
No errors now. And even I'm getting a "There are x orders pending sync!" message in the "Custom data stores" tab, which I didn't get before. Thanks!

**Reply by vedanshujain** [reply]
Sounds good, this is what you should see. Thanks for checking 🙌

**Reply by mrleemon** [reply]
@vedanshujain Since my last post, the "There are x orders pending sync!" message doesn't go away. It's stuck at "20 orders pending" forever:

![Screenshot 2022-09-26 at 15-29-36 WooCommerce settings - WordPress](https://user-images.githubusercontent.com/2060659/192290454-a2a8bbbc-4054-4475-b58d-6dfcb631f67e.png)

**Reply by piinthecloud** [reply]
Hi @mrleemon. Someone else had a similar issue in the [slack](https://woocommercecommunity.slack.com/archives/C043X91E72M/p1664193891267839). Can you try to click "Save changes"? This should re-trigger the sync. 

If that doesn't do the trick, try to manually sync the orders using the CLI. According to [this guide](https://developer.woocommerce.com/2022/09/16/high-performance-order-storage-call-for-testing-vol-2/#set-up-hpos), you should be able to speed up the initial synchronization by executing the sync in the WP CLI environment by running `wp wc cot sync`

**Reply by mrleemon** [reply]
Tried the two things:

- "Save changes" -> Nothing changes
- "wp wc cot sync" -> I get the following message:

```
There are 20 orders to be synced.
Order Data Sync  100% [====================================================================================] 0:00 / 0:00
Error: An infinite import loop has been detected, aborting.
```

**Reply by mrleemon** [reply]
Tried the same steps with the Hyyan WooCommerce Polylang Integration plugin disabled with the same results.

**Reply by vedanshujain** [reply]
Thanks for testing out thus far, seems like this is happening because we are not able to migrate alteast one order properly. I working on another build with more logging, but it will take a little bit.

**Reply by mrleemon** [reply]
I've updated to `7.0.0-hpos.dev.4` and now I no longer see the "There are 20 orders to be synced." message.
Also, I've run `wp wc cot sync` and now I'm getting the following:

```
There are 0 orders to be synced.
Warning: There are no orders to sync, aborting.
```

So, it seems this version fixed all the issues.

Thanks!

**Comment by mrleemon** [comment]
Just a question. The Hyyan WooCommerce Polylang Integration plugin uses the following hooks in its `Order.php` file:

```
woocommerce_checkout_update_order_meta
woocommerce_my_account_my_orders_query
woocommerce_order_item_product
woocommerce_order_data_store_cpt_get_orders_query

```

Are they still valid to use when using the new HPOS tables? Or are any of them deprecated?

Thanks

**Reply by vedanshujain** [reply]
They all should work as expected, except for the `woocommerce_order_data_store_cpt_get_orders_query` which won't be applied in case of HPOS. Instead, the recommendation is to modify the query while passing the parameters. (In case this is not possible, please share the use case or code example, we will add appropriate filters/hooks to have similar extensibility).

**Comment by mrleemon** [comment]
Another question, the Hyyan WooCommerce Polylang Integration plugin stores the language in which an order is made using a taxonomy named `language` which has as many terms as there are languages set up in the site. This is the Polylang function that does this task:

https://github.com/polylang/polylang/blob/a5b2aaec7682351a6f628a173267c3a371c3bb97/include/translated-post.php#L49

With HPOS tables, no posts (and no taxonomies) are used anywhere, so how can one replicate this functionality for HPOS orders?

Thanks

**Reply by vedanshujain** [reply]
We do insert a placeholder post in the posts table with same ID as order ID to maintain compatibility with terms and wp_comments. This method should continue to work as before.

**Reply by vedanshujain** [reply]
See more details about placeholder posts at https://developer.woocommerce.com/2022/09/29/high-performance-order-storage-backward-compatibility-and-synchronization/

**Reply by mrleemon** [reply]
So, is this placeholder post always created independently of the use of the HPOS tables?
Instead of referencing the old `shop_order` CPT, should the Hyyan plugin be referencing this new `shop_order_placehold` CPT to make use of related taxonomies and comments?
Is this a temporary feature to be deprecated in the near future? Or is it permanent?
